### PR TITLE
cloudflare[minor]: Add tool calling support

### PR DESCRIPTION
**Pull request by bracesproul**
todo: docs

### CODE REVIEW

File: examples/src/models/chat/integration_cloudflare_tools.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  console.log("chunk: ", chunk.content);
+  if (!finalChunk) {
+    finalChunk = chunk;
+  } else {
+    finalChunk = finalChunk.concat(chunk);
+  }
+}
+
+/*
+chunk:  <
```
Comment (jacoblee93) [inline examples/src/models/chat/integration_cloudflare_tools.ts:59]: Let's not do this - if too difficult to parse, let's just do what we did for Anthropic and delegate to `.invoke()`

Comment (jacoblee93) [reply examples/src/models/chat/integration_cloudflare_tools.ts:59]: This needs to populate `tool_call_chunks`

File: libs/langchain-cloudflare/src/chat_models.ts [hunk 2]
```diff
@@ -135,6 +224,20 @@ export class ChatCloudflareWorkersAI
     }
   }
 
+  override bindTools(
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/chat_models.ts:227]: Did we ever get around to that generic with structured output method?

File: libs/langchain-cloudflare/src/chat_models.ts [hunk 3]
```diff
@@ -3,25 +3,109 @@ import {
...
+    return (tools as ToolDefinition[]).map((tc) => ({
+      name: tc.function.name,
+      description:
+        tc.function.description ?? `A function named ${tc.function.name}.`,
+      parameters: tc.function.parameters,
+    }));
+  }
+  if (
+    tools.every((t) =>
+      Boolean(
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/chat_models.ts:83]: This constructor has some funny edge cases IIRC - let's just use `!!` to get truthy values only if we must

File: libs/langchain-cloudflare/src/chat_models.ts [hunk 4]
```diff
@@ -3,25 +3,109 @@ import {
...
-import type { CloudflareWorkersAIInput } from "./llms.js";
+import { StructuredToolInterface } from "@langchain/core/tools";
+import { Runnable } from "@langchain/core/runnables";
+import { isStructuredTool } from "@langchain/core/utils/function_calling";
+import { zodToJsonSchema } from "zod-to-json-schema";
+import { ToolCall } from "@langchain/core/messages/tool";
 import { convertEventStreamToIterableReadableDataStream } from "./utils/event_source_parse.js";
+import type { CloudflareWorkersAIInput } from "./llms.js";
+
+type CloudflareTool = {
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/chat_models.ts:29]: Do we need to export this type?

File: libs/langchain-cloudflare/src/chat_models.ts [hunk 5]
```diff
@@ -109,9 +193,14 @@ export class ChatCloudflareWorkersAI
   /**
    * Get the parameters used to invoke the model
    */
-  invocationParams(_options?: this["ParsedCallOptions"]) {
+  invocationParams(options?: this["ParsedCallOptions"]) {
+    const tools =
+      options?.tools && options.tools.length
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/chat_models.ts:198]: Passing `tools: undefined` is fine right?

File: libs/langchain-cloudflare/src/embeddings.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { Ai } from "@cloudflare/ai";
+import { Ai } from "@cloudflare/ai/dist/ai.js";
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/embeddings.ts:1]: Por que?

File: libs/langchain-cloudflare/src/chat_models.ts [hunk 7]
```diff
@@ -181,16 +284,55 @@ export class ChatCloudflareWorkersAI
...
         });
         yield generationChunk;
         // eslint-disable-next-line no-void
         void runManager?.handleLLMNewToken(generationChunk.text ?? "");
+        fullMessage += parsedChunk.response;
+      }
+    }
+
+    if (
+      fullMessage.startsWith(TOOL_CALL_START_TOKEN) &&
```
Comment (jacoblee93) [inline libs/langchain-cloudflare/src/chat_models.ts:304]: This is consistent for all models?

File: docs/core_docs/docs/integrations/chat/cloudflare_workersai.mdx [hunk 8]
```diff
@@ -26,3 +26,15 @@ import UnifiedModelParamsTooltip from "@mdx_components/unified_model_params_tool
...
+
+### Tool calling
+
+:::note
+Tool calling is only available in `@langchain/cloudflare` version `0.0.7` and above.
+:::
+
+Cloudflare's API now supports tool calling! The below example demonstrates how to invoke and stream tool calls.
+
+import ToolCalling from "@examples/models/chat/integration_cloudflare_tools.ts";
```
Comment (jacoblee93) [inline docs/core_docs/docs/integrations/chat/cloudflare_workersai.mdx:38]: Is it only specific models?

File: libs/langchain-cloudflare/src/chat_models.ts (hunk 2 above)
Comment (bracesproul) [reply libs/langchain-cloudflare/src/chat_models.ts:227]: Not merged on PY side yet. Pending review

File: libs/langchain-cloudflare/src/chat_models.ts (hunk 4 above)
Comment (bracesproul) [reply libs/langchain-cloudflare/src/chat_models.ts:29]: good call, prob should

File: libs/langchain-cloudflare/src/chat_models.ts (hunk 5 above)
Comment (bracesproul) [reply libs/langchain-cloudflare/src/chat_models.ts:198]: yes

File: libs/langchain-cloudflare/src/chat_models.ts (hunk 7 above)
Comment (bracesproul) [reply libs/langchain-cloudflare/src/chat_models.ts:304]: dropped in favor of `.invoke`

File: docs/core_docs/docs/integrations/chat/cloudflare_workersai.mdx (hunk 8 above)
Comment (bracesproul) [reply docs/core_docs/docs/integrations/chat/cloudflare_workersai.mdx:38]: Will confirm w/ team and add a note.

File: libs/langchain-cloudflare/src/embeddings.ts (hunk 6 above)
Comment (bracesproul) [reply libs/langchain-cloudflare/src/embeddings.ts:1]: I was unable to run in `examples`, or in a local project due to this error:
```
file:///Users/bracesproul/code/lang-chain-ai/wt/brace/cloudflare-tools/libs/langchain-cloudflare/dist/embeddings.js:2
import { Ai } from "@cloudflare/ai";
         ^^
SyntaxError: The requested module '@cloudflare/ai' does not provide an export named 'Ai'
```
When I switched the import path the error went away.

Comment (bracesproul) [reply libs/langchain-cloudflare/src/embeddings.ts:1]: awe shit, in triggering this error again to show you, it's now persisting even with that new path. Will try and fix

Comment (craigsdennis) [reply libs/langchain-cloudflare/src/embeddings.ts]: That package is on the way out! 

There is a [TypeScript SDK](https://github.com/cloudflare/cloudflare-typescript) now that wraps nicely...but also it's on the Binding `env.AI.run` not sure how to do this.

Comment (Cherry) [reply libs/langchain-cloudflare/src/embeddings.ts]: I have a PR up that uses the native binding: https://github.com/langchain-ai/langchainjs/pull/5287

**Comment by bracesproul** [comment]
Closing: stale

**Comment by tjnangosha** [comment]
Hi team, @bracesproul,

I am wondering why this was closed.
I am currently  using `"@langchain/cloudflare": "^0.1.0",` 

and I would like to know if something like
```
const { CloudflareWorkersAI,  } = require('@langchain/cloudflare');
const llm = new CloudflareWorkersAI({cloudflareAccountId, cloudflareApiToken});
llm.bindToolls() // I can't seem to do this at the moment
```

is possible in javascript

Atm, I am mostly doing something like `tool.invoke()` and I don't know if manually invoking tools is good practice

**Pull request closed without merging**